File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    dfs(leftMin, root, rightMax){
        if(!root){
            return true
        }
        if(!(leftMin < root.val && root.val < rightMax)){
            return false
        }
        return this.dfs(leftMin, root.left, root.val) && this.dfs(root.val, root.right, rightMax) 
    }
    isValidBST(root) {
        return this.dfs(-10001, root, 10001)
    }
}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        max_depth = float("-inf")

        if(not(root)):
            return 0

        def dfs(node, lvl):
            nonlocal max_depth
            if(not(node)):
                return 0
            
            max_depth = max(max_depth, lvl)
            dfs(node.left, lvl + 1)
            dfs(node.right, lvl + 1)
            

        dfs(root, 1)
        return max_depth
